VERSIONING.md

```diff
@@ -0,0 +1,103 @@
+[Obsolete](https://docs.microsoft.com/dotnet/api/system.obsoleteattribute)
+first.
+
+## Pre-releases
+
+Pre-release packages are identified by identifies such as -Alpha, -Beta, -RC
+etc. There is no API guarantees in pre-releases. In general, an RC pre-release
+is more stable than a Beta release, which is more stable than an Alpha release.
+
+### Public API change detection
+
+For convenience, every project in this repo uses
+[PublicApiAnalyzers](https://github.com/dotnet/roslyn-analyzers/tree/master/src/PublicApiAnalyzers)
+and lists public API in the directory "publicAPI". Any changes to public API,
+without corresponding changes here will result in build breaks - this helps
+catch any unintended changes to public API from being shipped accidentally. This
+also helps reviewers quickly understand if a given PR is making public API
+changes. For example,
+[this](https://github.com/open-telemetry/opentelemetry-dotnet/tree/master/src/OpenTelemetry.Instrumentation.AspNetCore/.publicApi)
+shows the public per target framework for the
+`OpenTelemetry.Instrumentation.AspNetCore` package.
+
+Since no stable version has been released so far, every API is listed in the
+"Unshipped.txt" file. Once 1.0.0 is shipped, it'll be moved to "Shipped.txt"
+file.
+
+## Packaging
+
+OpenTelemetry is structured around signals like Traces, Metrics, Logs, Baggage
+etc. OpenTelemetry .NET does not ship separate package per signal. There is a
+single package which includes all the signals. i.e `OpenTelemetry.API` package
+will consist of API components from *all* the signals. `OpenTelemetry` package
+will consist of SDK components from *all* the signals. Instrumentations also
+follow the same model - for example, There will be a single package
+`OpenTelemetry.Instrumentation.AspNetCore` for ASP.NET Core instrumentation,
+which produces Traces, Metrics and handles propagation, instead of separate
+packages for traces and metrics.
+
+## Experimental Signals
+
+### Current approach
+
+Given the fact that all signals are shipped as a single package, the following
+approach is currently used to deal with experimental signals.
+
+Any experimental signal which is shipped as part of a normal package will be
+marked
+[Obsolete](https://docs.microsoft.com/dotnet/api/system.obsoleteattribute).
+(Ideally we need an "Experimental" attribute, but .NET has only built-in support
+for "Obsolete", so we are leveraging it. Using "Obsolete" code results in
+compile time warnings to caution users.) This allows an experimental signal to
+co-exist in the same package as other, non-experimental signals. For example,
+the OpenTelemetry 1.0.0 package will consist of Traces and Metrics. As Metrics
+signal is not ready for stable release, every method dealing with Metrics will
+be marked "Obsolete". Once Metrics signal gets stable, it'll be added to a
+release with minor version bump, say 1.2.0. The "Obsolete" Metric methods can be
+removed with a major version bump.
+
+### Alternate option
```

Comment:
I'm not settled on this yet, but I'm beginning to warm up to this approach. My main concern was that early adopters of metrics will likely strive to stay current as metrics functionality evolves. As metrics becomes more and more stable, so will the application code they have written. I think it would be ideal that when metrics goes GA, early adopters would not then be required to switch their code to use a different package/namespace.

Two questions:
* If we initially shipped metrics in a `OpenTelemetry.Metrics.Experimental` package, could we simply use the `OpenTelemetry.Metrics` namespace?
* Perhaps the separate package approach + the use of the obsolete attribute would be beneficial? That is, the build warnings are a nice signal that you're using something at your own risk.

Reply:
Assuming we get metrics into own separate package,  yes agree we should use same namespace. The goal would be - early adopters who were using the experimental, will find it easy to migrate. The migration involves just removing the metric-experimental package, and upgrading the opentelemetry package to the one containing metrics, no code change. (Assuming they kept up-to-date with the experimental).

Reply:
If we are using separate package approach, then we can name it simply OpenTelemetry.API/SDK version 1.0.0-beta.1. The pre-release version already indicate its a unstable package, so additionally marking obsolete won't be necessary.

Reply:
> If we initially shipped metrics in a OpenTelemetry.Metrics.Experimental package, could we simply use the OpenTelemetry.Metrics namespace for all public API?

If we go with separate packages, I'm all up for this (definitely making the users' lives easier).

Reply:
> then we can name it simply OpenTelemetry.API/SDK version 1.0.0-beta.1.

You mean the same package name as we have today? That is `OpenTelemetry` and `OpenTelemetry.Api`?

Since we have already released version 1.0.0-rc1.1 this won't work because 1.0.0-rc1.1 > 1.0.0-beta.1 based on how semver and nuget handles things.

I envisioned that separate package would mean difference package name, so something like `OpenTelemetry.Metrics.Experimental` as you suggested.

Reply:
Yea you are right. I was trying to say - we can drop the word "Experimental" from the package name. Like OpenTelemetry.Metrics .9.0-alpha1 for instance? the pre-release alone is sufficient to indicate its not stable. right?